File: TS/binarySearchTree.ts
export class BSTNode<T> {
  value: T
  left: BSTNode<T> | null = null
  right: BSTNode<T> | null = null

  constructor(value: T) {
    this.value = value
  }
}

export class BinarySearchTree<T> {
  private root: BSTNode<T> | null = null
  private compareFn: (a: T, b: T) => number

  constructor(compareFn?: (a: T, b: T) => number) {
    this.compareFn = compareFn ?? ((a: T, b: T) => a < b ? -1 : a > b ? 1 : 0)
  }

  insert(value: T): void {
    this.root = this.insertNode(this.root, value)
  }

  private insertNode(node: BSTNode<T> | null, value: T): BSTNode<T> {
    if (!node) return new BSTNode(value)
    const cmp = this.compareFn(value, node.value)
    if (cmp < 0) node.left = this.insertNode(node.left, value)
    else if (cmp > 0) node.right = this.insertNode(node.right, value)
    return node
  }

  search(value: T): boolean {
    let current = this.root
    while (current) {
      const cmp = this.compareFn(value, current.value)
      if (cmp === 0) return true
      current = cmp < 0 ? current.left : current.right
    }
    return false
  }

  delete(value: T): void {
    this.root = this.deleteNode(this.root, value)
  }

  private deleteNode(node: BSTNode<T> | null, value: T): BSTNode<T> | null {
    if (!node) return null
    const cmp = this.compareFn(value, node.value)
    if (cmp < 0) { node.left = this.deleteNode(node.left, value); return node }
    if (cmp > 0) { node.right = this.deleteNode(node.right, value); return node }

    if (!node.left) return node.right
    if (!node.right) return node.left

    const minNode = this.findMin(node.right)
    node.value = minNode.value
    node.right = this.deleteNode(node.right, minNode.value)
    return node
  }

  private findMin(node: BSTNode<T>): BSTNode<T> {
    while (node.left) node = node.left
    return node
  }

  inOrder(): T[] {
    const result: T[] = []
    this.inOrderTraverse(this.root, result)
    return result
  }

  private inOrderTraverse(node: BSTNode<T> | null, result: T[]): void {
    if (!node) return
    this.inOrderTraverse(node.left, result)
    result.push(node.value)
    this.inOrderTraverse(node.right, result)
  }

  preOrder(): T[] {
    const result: T[] = []
    this.preOrderTraverse(this.root, result)
    return result
  }

  private preOrderTraverse(node: BSTNode<T> | null, result: T[]): void {
    if (!node) return
    result.push(node.value)
    this.preOrderTraverse(node.left, result)
    this.preOrderTraverse(node.right, result)
  }

  postOrder(): T[] {
    const result: T[] = []
    this.postOrderTraverse(this.root, result)
    return result
  }

  private postOrderTraverse(node: BSTNode<T> | null, result: T[]): void {
    if (!node) return
    this.postOrderTraverse(node.left, result)
    this.postOrderTraverse(node.right, result)
    result.push(node.value)
  }

  min(): T | null {
    if (!this.root) return null
    let current = this.root
    while (current.left) current = current.left
    return current.value
  }

  max(): T | null {
    if (!this.root) return null
    let current = this.root
    while (current.right) current = current.right
    return current.value
  }
}

File: CSHARP/BinarySearchTree.cs
using System;
using System.Collections.Generic;

namespace MtplSample
{
    public class BinarySearchTree<T> where T : IComparable<T>
    {
        private TreeNode<T> _root;

        public int Count { get; private set; }

        public void Insert(T value)
        {
            _root = InsertNode(_root, value);
            Count++;
        }

        private TreeNode<T> InsertNode(TreeNode<T> node, T value)
        {
            if (node == null) return new TreeNode<T>(value);

            if (value.CompareTo(node.Value) < 0)
                node.Left = InsertNode(node.Left, value);
            else if (value.CompareTo(node.Value) > 0)
                node.Right = InsertNode(node.Right, value);
            else
                Count--;
            return node;
        }

        public bool Search(T value)
        {
            return SearchNode(_root, value) != null;
        }

        private TreeNode<T> SearchNode(TreeNode<T> node, T value)
        {
            if (node == null) return null;
            if (value.CompareTo(node.Value) == 0) return node;
            if (value.CompareTo(node.Value) < 0)
                return SearchNode(node.Left, value);
            return SearchNode(node.Right, value);
        }

        public void Delete(T value)
        {
            _root = DeleteNode(_root, value);
        }

        private TreeNode<T> DeleteNode(TreeNode<T> node, T value)
        {
            if (node == null) return null;

            if (value.CompareTo(node.Value) < 0)
                node.Left = DeleteNode(node.Left, value);
            else if (value.CompareTo(node.Value) > 0)
                node.Right = DeleteNode(node.Right, value);
            else
            {
                if (node.Left == null) return node.Right;
                if (node.Right == null) return node.Left;

                var minNode = FindMin(node.Right);
                node.Value = minNode.Value;
                node.Right = DeleteNode(node.Right, minNode.Value);
            }
            return node;
        }

        private TreeNode<T> FindMin(TreeNode<T> node)
        {
            while (node.Left != null) node = node.Left;
            return node;
        }

        public List<T> InOrderTraversal()
        {
            var result = new List<T>();
            InOrder(_root, result);
            return result;
        }

        private void InOrder(TreeNode<T> node, List<T> result)
        {
            if (node == null) return;
            InOrder(node.Left, result);
            result.Add(node.Value);
            InOrder(node.Right, result);
        }
    }

    public class TreeNode<T>
    {
        public T Value { get; set; }
        public TreeNode<T> Left { get; set; }
        public TreeNode<T> Right { get; set; }

        public TreeNode(T value)
        {
            Value = value;
        }
    }
}
